#Prediction of Diabetes using Bagging technique

Import libraries

In [5]:
import keras
import numpy as np
import pandas as pd

Import dataset

In [6]:
df = pd.read_csv('diabetes.csv')

In [7]:
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

checking for null values

In [120]:
df.isna().sum()     #no null values

Index                       0
Age                         0
Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Outcome                     0
dtype: int64

There are no null values

In [121]:
df.size

7680

In [9]:
rows = len(df)
rows

768

In [10]:
df.index

RangeIndex(start=0, stop=768, step=1)

In [11]:
df['Index'] = df.index + 1

In [12]:
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  Index  
0                     0.627   50        1      1  
1                     0.351   31        0      2  
2                     0.672   32        1      3  
3                     0.167   21        0      4  
4                     2.288   33        1      5

rearranging columns

In [13]:
df=df.reindex(columns=['Index','Age','Pregnancies','Glucose','BloodPressure','SkinThickness','Insulin','BMI','DiabetesPedigreeFunction',"Outcome"])

In [163]:
df.head()

Index  Age  Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin  \
0      1   50            6      148             72             35        0   
1      2   31            1       85             66             29        0   
2      3   32            8      183             64              0        0   
3      4   21            1       89             66             23       94   
4      5   33            0      137             40             35      168   

    BMI  DiabetesPedigreeFunction  Outcome  
0  33.6                     0.627        1  
1  26.6                     0.351        0  
2  23.3                     0.672        1  
3  28.1                     0.167        0  
4  43.1                     2.288        1

In [15]:
part1 = df.sample(frac=0.5)

In [16]:
part1

Index  Age  Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin  \
128    129   40            1      117             88             24      145   
669    670   45            9      154             78             30      100   
76      77   41            7       62             78              0        0   
71      72   26            5      139             64             35      140   
360    361   29            5      189             64             33      325   
..     ...  ...          ...      ...            ...            ...      ...   
202    203   32            0      108             68             20        0   
651    652   27            1      117             60             23      106   
656    657   22            2      101             58             35       90   
52      53   30            5       88             66             21       23   
556    557   30            1       97             70             40        0   

      BMI  DiabetesPedigreeFunction  Outcome  
128  34.5                     0.403        1  
669  30.9                     0.164        0  
76   32.6                     0.391        0  
71   28.6                     0.411        0  
360  31.2                     0.583        1  
..    ...                       ...      ...  
202  27.3                     0.787        0  
651  33.8                     0.466        0  
656  21.8                     0.155        0  
52   24.4                     0.342        0  
556  38.1                     0.218        0  

[384 rows x 10 columns]

In [17]:
len(part1)
# no of rows = 768 --> n/2 = 384

384

In [18]:
# rest of the part
part2 = df.drop(part1.index)

In [20]:
len(part2)

384

checking if correctly divided and no repeating rows

In [23]:
print(part1.loc[part1['Index']==17])

    Index  Age  Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin  \
16     17   31            0      118             84             47      230   

     BMI  DiabetesPedigreeFunction  Outcome  
16  45.8                     0.551        1  


In [24]:
print(part2.loc[part2['Index']==17])

Empty DataFrame
Columns: [Index, Age, Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, Outcome]
Index: []


#Using decision tree classifier with part1

splitting into X and y

In [75]:
X1 = part1.iloc[:, :-1].values
# X1 = pd.DataFrame(X1)
X1

array([[1.29e+02, 4.00e+01, 1.00e+00, ..., 1.45e+02, 3.45e+01, 4.03e-01],
       [6.70e+02, 4.50e+01, 9.00e+00, ..., 1.00e+02, 3.09e+01, 1.64e-01],
       [7.70e+01, 4.10e+01, 7.00e+00, ..., 0.00e+00, 3.26e+01, 3.91e-01],
       ...,
       [6.57e+02, 2.20e+01, 2.00e+00, ..., 9.00e+01, 2.18e+01, 1.55e-01],
       [5.30e+01, 3.00e+01, 5.00e+00, ..., 2.30e+01, 2.44e+01, 3.42e-01],
       [5.57e+02, 3.00e+01, 1.00e+00, ..., 0.00e+00, 3.81e+01, 2.18e-01]])

In [26]:
y1 = part1.iloc[:, -1].values


In [76]:
from sklearn import tree
dt = tree.DecisionTreeClassifier()

In [77]:
dt = dt.fit(X1,y1)

In [78]:
 y1_pred = dt.predict(X1)

In [79]:
y1_pred

array([1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0,
       1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1,
       0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1,
       0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0,
       1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0,
       1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,

In [80]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y1, y1_pred)
print(cm)
accuracy_score(y1, y1_pred)


[[267   0]
 [  0 117]]


1.0

#Using knn classifier with part2

In [45]:
X2 = part2.iloc[:, :-1].values
X2

array([[3.00e+00, 3.20e+01, 8.00e+00, ..., 0.00e+00, 2.33e+01, 6.72e-01],
       [7.00e+00, 2.60e+01, 3.00e+00, ..., 8.80e+01, 3.10e+01, 2.48e-01],
       [8.00e+00, 2.90e+01, 1.00e+01, ..., 0.00e+00, 3.53e+01, 1.34e-01],
       ...,
       [7.65e+02, 2.70e+01, 2.00e+00, ..., 0.00e+00, 3.68e+01, 3.40e-01],
       [7.66e+02, 3.00e+01, 5.00e+00, ..., 1.12e+02, 2.62e+01, 2.45e-01],
       [7.67e+02, 4.70e+01, 1.00e+00, ..., 0.00e+00, 3.01e+01, 3.49e-01]])

In [46]:
y2 = part2.iloc[:, -1].values


knn

In [47]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X2 = sc.fit_transform(X2)


In [48]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
knn.fit(X2, y2)

KNeighborsClassifier()

In [49]:
y2_pred = knn.predict(X2)

In [50]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm2 = confusion_matrix(y2, y2_pred)
print(cm2)
accuracy_score(y2, y2_pred)

[[195  38]
 [ 43 108]]


0.7890625

Variation in accuracy

part1 - Decision tree accuracy 100%

part2 - KNN accuracy 79%

#Bagging

In [51]:
from sklearn.pipeline import Pipeline

pipeline = Pipeline([
    ('DT', dt),
    ('KNN',knn)
])
pipeline

Pipeline(steps=[('DT', DecisionTreeClassifier()),
                ('KNN', KNeighborsClassifier())])

Taking random 10% from data as  test samples

In [156]:
testdf = df.sample(frac=0.1)
len(testdf)

77

In [157]:
testdf    #random

Index  Age  Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin  \
673    674   22            3      123            100             35      240   
325    326   24            1      157             72             21      168   
699    700   26            4      118             70              0        0   
118    119   22            4       97             60             23        0   
33      34   28            6       92             92              0        0   
..     ...  ...          ...      ...            ...            ...      ...   
487    488   58            0      173             78             32      265   
101    102   22            1      151             60              0        0   
132    133   30            3      170             64             37      225   
17      18   31            7      107             74              0        0   
12      13   57           10      139             80              0        0   

      BMI  DiabetesPedigreeFunction  Outcome  
673  57.3                     0.880        0  
325  25.6                     0.123        0  
699  44.5                     0.904        0  
118  28.2                     0.443        0  
33   19.9                     0.188        0  
..    ...                       ...      ...  
487  46.5                     1.159        0  
101  26.1                     0.179        0  
132  34.5                     0.356        1  
17   29.6                     0.254        1  
12   27.1                     1.441        0  

[77 rows x 10 columns]

In [158]:
X = testdf.iloc[:, :-1].values
y = testdf.iloc[:, -1].values

making prediction on sample data using pretrained models

In [159]:
dt_pred = dt.predict(X)
knn_pred = knn.predict(X)

In [164]:
# performing bagging
avg_pred = (dt_pred + knn_pred) / 2

In [165]:
# converting to binary
bag_pred = []
for i in avg_pred:
    if i > 0.5:
        bag_pred.append(1)
    else:
        bag_pred.append(0)

In [167]:
from sklearn.metrics import confusion_matrix, accuracy_score
# Assuming y_test contains the true labels

cm4 = confusion_matrix(y , bag_pred)
print(cm4)
accuracy4 = accuracy_score(y, bag_pred)

print("Accuracy:", accuracy4)

[[45  7]
 [ 3 22]]
Accuracy: 0.8701298701298701


part1 - Decision tree accuracy 100%  (predicted on same dataset used for training)

part2 - KNN accuracy 81.5%

after bagging  accuracy 87.01%  (accuracy has increased)

Trying Decision tree after test and train splitting



In [114]:
X3 =df.iloc[: , :-1].values
y3 = df.iloc[: , -1].values

In [115]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(  X3, y3, test_size=0.33, random_state=42)

In [116]:
dt1=dt.fit(X_train,y_train)

In [117]:
y_pred3 = dt1.predict(X_test)

In [118]:
accuracy = accuracy_score(y_test, y_pred3)

print("Accuracy:", accuracy)

Accuracy: 0.7086614173228346


we see that decision tree performs very well when tested on same dataset , but it has poor accuracy score with unseen data.